Patch by MP-Greg
installer.rb - Windows allows paths to use '/' for a path separator, adjust code
to allow for it
test_case.rb - code so it can find test directory from $LOAD_PATH
diff --git a/lib/rubygems/installer.rb b/lib/rubygems/installer.rb
index 8cb89e2e40..2e803be182 100644
--- a/lib/rubygems/installer.rb
+++ b/lib/rubygems/installer.rb
@@ -698,10 +698,12 @@ def process_options # :nodoc:
 
   def check_that_user_bin_dir_is_in_path # :nodoc:
     user_bin_dir = @bin_dir || Gem.bindir(gem_home)
-    user_bin_dir = user_bin_dir.gsub(File::SEPARATOR, File::ALT_SEPARATOR) if
-      File::ALT_SEPARATOR
-
     path = ENV['PATH']
+    if File::ALT_SEPARATOR
+      user_bin_dir = user_bin_dir.gsub(File::SEPARATOR, File::ALT_SEPARATOR)
+      path = path.gsub(File::SEPARATOR, File::ALT_SEPARATOR)
+    end
+    
     if Gem.win_platform? then
       path = path.downcase
       user_bin_dir = user_bin_dir.downcase
diff --git a/lib/rubygems/test_case.rb b/lib/rubygems/test_case.rb
index ced33c4d11..1f28a1666b 100644
--- a/lib/rubygems/test_case.rb
+++ b/lib/rubygems/test_case.rb
@@ -1320,8 +1320,6 @@ def self.rubybin
   end
 
   @@ruby = rubybin
-  @@good_rake = "#{rubybin} \"#{File.expand_path('../../../test/rubygems/good_rake.rb', __FILE__)}\""
-  @@bad_rake = "#{rubybin} \"#{File.expand_path('../../../test/rubygems/bad_rake.rb', __FILE__)}\""
 
   ##
   # Construct a new Gem::Dependency.
@@ -1509,14 +1507,10 @@ def self.load_cert cert_name
 
   def self.cert_path cert_name
     if 32 == (Time.at(2**32) rescue 32) then
-      cert_file =
-        File.expand_path "../../../test/rubygems/#{cert_name}_cert_32.pem",
-                         __FILE__
-
+      cert_file = File.join test_path, "#{cert_name}_cert_32.pem"
       return cert_file if File.exist? cert_file
     end
-
-    File.expand_path "../../../test/rubygems/#{cert_name}_cert.pem", __FILE__
+    File.join test_path, "#{cert_name}_cert.pem"
   end
 
   ##
@@ -1534,7 +1528,23 @@ def self.load_key key_name, passphrase = nil
   # Returns the path to the key named +key_name+ from <tt>test/rubygems</tt>
 
   def self.key_path key_name
-    File.expand_path "../../../test/rubygems/#{key_name}_key.pem", __FILE__
+    File.join test_path, "#{key_name}_key.pem"
+  end
+
+  def self.test_path
+    @@test_path ||= begin
+      temp = $LOAD_PATH.reverse.find { |p| /test\/(lib|rubygems|-ext-)\z/ =~ p } || ''
+      if temp.empty?
+        temp = File.expand_path('../../../test/rubygems', __FILE__)
+      else
+        temp = File.join File.dirname(temp), 'rubygems'
+        temp = '' unless File.exist? File.join(temp, 'test_gem.rb')
+      end
+      r_bin = Gem::TestCase.rubybin
+      @@good_rake ||= "#{r_bin} \"#{File.join temp, 'good_rake.rb'}\""
+      @@bad_rake  ||= "#{r_bin} \"#{File.join temp, 'bad_rake.rb'}\""
+      temp
+    end
   end
 
   # :stopdoc:
@@ -1596,3 +1606,5 @@ def self.key_path key_name
 END {tmpdirs.each {|dir| Dir.rmdir(dir)} if $$ == pid}
 Gem.clear_paths
 Gem.loaded_specs.clear
+# $LOAD_PATH has path to tests loaded
+Gem::TestCase.test_path
